README.md
SMSJ
====

[![Build Status](https://travis-ci.org/marre/smsj.png)](https://travis-ci.org/marre/smsj)

Move in progress from sourceforge.

http://smsj.sourceforge.net/

The first smsj checkin was done at Mar 15, 2002

View Changelog.txt to release notes

Changelog.txt
Version 0.1.2
	UcpTransport:
		- Añadido TCP timeout
	SmsOtaMessage
		- Soporte a mensajes OTA
		
Version 0.1.1
	SmsTextMessage:
		- Autodetect encoding and alphabet
		- Latin1 coding fixed
		- GSM message normalized
		- Long message type accepted (to send message with 160 characters or more)

	UcpSeries50:
		- UDH fixed (allow multi-part messages)

	UcpTransport:
		- Field OTOA fixed
		- Latin1 without UDH accepted
		- Send method return the last result (improve)


src/main/java/org/marre/util/StringUtil.java
/* ***** BEGIN LICENSE BLOCK *****
 * Version: MPL 1.1/GPL 2.0/LGPL 2.1
 *
 * The contents of this file are subject to the Mozilla Public License Version
 * 1.1 (the "License"); you may not use this file except in compliance with
 * the License. You may obtain a copy of the License at
 * http://www.mozilla.org/MPL/
 *
 * Software distributed under the License is distributed on an "AS IS" basis,
 * WITHOUT WARRANTY OF ANY KIND, either express or implied. See the License
 * for the specific language governing rights and limitations under the
 * License.
 *
 * The Original Code is "SMS Library for the Java platform".
 *
 * The Initial Developer of the Original Code is Markus Eriksson.
 * Portions created by the Initial Developer are Copyright (C) 2002
 * the Initial Developer. All Rights Reserved.
 *
 * Contributor(s):
 *
 * Alternatively, the contents of this file may be used under the terms of
 * either the GNU General Public License Version 2 or later (the "GPL"), or
 * the GNU Lesser General Public License Version 2.1 or later (the "LGPL"),
 * in which case the provisions of the GPL or the LGPL are applicable instead
 * of those above. If you wish to allow use of your version of this file only
 * under the terms of either the GPL or the LGPL, and not to allow others to
 * use your version of this file under the terms of the MPL, indicate your
 * decision by deleting the provisions above and replace them with the notice
 * and other provisions required by the GPL or the LGPL. If you do not delete
 * the provisions above, a recipient may use your version of this file under
 * the terms of any one of the MPL, the GPL or the LGPL.
 *
 * ***** END LICENSE BLOCK ***** */
package org.marre.util;

import java.util.Random;

/**
 * Various functions to encode and decode strings.
 * 
 * @author Markus Eriksson
 */
public final class StringUtil
{
	private static final char[] RANDOM_CHARS = "abcdefghijklmnopqrstuvwxyz1234567890".toCharArray();
	private static final Random rnd_ = new Random();

	/**
	 * This class isn't intended to be instantiated.
	 */
	private StringUtil()
	{
	}

	/**
	 * 
	 * @param stringTable
	 * @param text
	 * @return
	 */
	public static int findString(String[] stringTable, String text)
	{
		if (stringTable != null)
		{
			for (int i = 0; i < stringTable.length; i++)
			{
				if ((stringTable[i] != null) && (stringTable[i].equals(text)))
				{
					return i;
				}
			}
		}
		return -1;
	}

	/**
	 * Converts a byte array to a string with hex values.
	 * 
	 * @param data
	 *            Data to convert
	 * @return the encoded string
	 */
	public static String bytesToHexString(byte[] data)
	{
		StringBuilder hexStrBuff = new StringBuilder(data.length * 2);
		for (byte aData : data)
		{
			String hexByteStr = Integer.toHexString(aData & 0xff).toUpperCase();
			if (hexByteStr.length() == 1)
			{
				hexStrBuff.append("0");
			}
			hexStrBuff.append(hexByteStr);
		}
		return hexStrBuff.toString();
	}

	/**
	 * Converts a byte to a string with hex values.
	 * 
	 * @param data
	 *            Byte to convert
	 * @return the encoded string
	 */
	public static String byteToHexString(byte data)
	{
		StringBuilder hexStrBuff = new StringBuilder(2);
		String hexByteStr = Integer.toHexString(data & 0xff).toUpperCase();
		if (hexByteStr.length() == 1)
		{
			hexStrBuff.append("0");
		}
		hexStrBuff.append(hexByteStr);
		return hexStrBuff.toString();
	}

	/**
	 * Converts a string of hex characters to a byte array.
	 * 
	 * @param hexString
	 *            The hex string to read
	 * @return the resulting byte array
	 */
	public static byte[] hexStringToBytes(String hexString)
	{
		byte[] data = new byte[hexString.length() / 2];
		for (int i = 0; i < data.length; i++)
		{
			String a = hexString.substring(i * 2, i * 2 + 2);
			data[i] = (byte) Integer.parseInt(a, 16);
		}
		return data;
	}

	/**
	 * Method intToString.
	 * 
	 * Converst an integer to nChars characters
	 * 
	 * @param value
	 *            Integer value
	 * @param nChars
	 *            Number of chars to represent the "value"
	 * @return String The string representing "value"
	 */
	public static String intToString(int value, int nChars)
	{
		String strValue = Integer.toString(value);
		StringBuilder strBuf = new StringBuilder(nChars);
		for (int i = strValue.length(); i < nChars; i++)
		{
			strBuf.append('0');
		}
		strBuf.append(strValue);
		return strBuf.toString();
	}

	/**
	 * Generates a random string of the given length.
	 * 
	 * "abcdefghijklmnopqrstuvwxyz1234567890"
	 * 
	 * @param length
	 * @return A random string
	 */
	public static String randString(int length)
	{
		StringBuilder sb = new StringBuilder(length);
		for (int i = 0; i < length; i++)
		{
			sb.append(RANDOM_CHARS[rnd_.nextInt(RANDOM_CHARS.length)]);
		}
		return sb.toString();
	}

	/**
	 * Try to detect the enconding text
	 * 
	 * @param text
	 * @return
	 */
	public static String detectEncodingType(String text)
	{
		String chars[] = { "€", "á", "é", "í", "ó", "ú", "ü", "ñ", "Á", "É", "Í", "Ó", "Ú", "Ü", "Ñ", "[" };
		String codings[] = { "Windows-1252", "ISO-8859-15", "UTF-8" };
		for (int i = 0; i < chars.length; i++)
		{
			for (int j = 0; j < codings.length; j++)
			{
				try
				{
					String coding = codings[j];
					String cChar = new String(chars[i].getBytes(), coding);
					if (text.contains(cChar))
						return coding;
				}
				catch (Exception ignore)
				{
				}
			}
		}
		return "US-ASCII";
	}

	public static boolean isNullOrEmpty(String str)
	{
		return null == str || str.length() == 0;
	}
}


src/main/java/org/marre/sms/SmsTextMessage.java
/* ***** BEGIN LICENSE BLOCK *****
 * Version: MPL 1.1/GPL 2.0/LGPL 2.1
 *
 * The contents of this file are subject to the Mozilla Public License Version
 * 1.1 (the "License"); you may not use this file except in compliance with
 * the License. You may obtain a copy of the License at
 * http://www.mozilla.org/MPL/
 *
 * Software distributed under the License is distributed on an "AS IS" basis,
 * WITHOUT WARRANTY OF ANY KIND, either express or implied. See the License
 * for the specific language governing rights and limitations under the
 * License.
 *
 * The Original Code is "SMS Library for the Java platform".
 *
 * The Initial Developer of the Original Code is Markus Eriksson.
 * Portions created by the Initial Developer are Copyright (C) 2002
 * the Initial Developer. All Rights Reserved.
 *
 * Contributor(s):
 *
 * Alternatively, the contents of this file may be used under the terms of
 * either the GNU General Public License Version 2 or later (the "GPL"), or
 * the GNU Lesser General Public License Version 2.1 or later (the "LGPL"),
 * in which case the provisions of the GPL or the LGPL are applicable instead
 * of those above. If you wish to allow use of your version of this file only
 * under the terms of either the GPL or the LGPL, and not to allow others to
 * use your version of this file under the terms of the MPL, indicate your
 * decision by deleting the provisions above and replace them with the notice
 * and other provisions required by the GPL or the LGPL. If you do not delete
 * the provisions above, a recipient may use your version of this file under
 * the terms of any one of the MPL, the GPL or the LGPL.
 *
 * ***** END LICENSE BLOCK ***** */
package org.marre.sms;

import java.io.UnsupportedEncodingException;
import java.text.Normalizer;

import org.marre.util.StringUtil;

/**
 * Represents a text message.
 * <p>
 * The text can be sent in unicode (max 70 chars/SMS), 8-bit (max 140 chars/SMS) or GSM encoding (max 160 chars/SMS).
 *
 * @author Markus Eriksson
 * @version $Id$
 */
public class SmsTextMessage extends SmsConcatMessage
{
	protected String text_;
	protected SmsDcs dcs_;

	/**
	 * Creates an SmsTextMessage with the given dcs.
	 * 
	 * @param msg
	 *            The message
	 * @param dcs
	 *            The data coding scheme
	 */
	public SmsTextMessage(String msg, SmsDcs dcs)
	{
		setText(msg, dcs);
	}

	/**
	 * Creates an SmsTextMessage with the given alphabet and message class.
	 *
	 * @param msg
	 *            The message
	 * @param alphabet
	 *            The alphabet
	 * @param messageClass
	 *            The messageclass
	 */
	public SmsTextMessage(String msg, SmsAlphabet alphabet, SmsMsgClass messageClass)
	{
		this(msg, alphabet, messageClass, false);
	}

	/**
	 * Creates an SmsTextMessage with the given alphabet and message class.
	 *
	 * @param msg
	 *            The message
	 * @param alphabet
	 *            The alphabet
	 * @param messageClass
	 *            The messageclass
	 */
	public SmsTextMessage(String msg, SmsAlphabet alphabet, SmsMsgClass messageClass, boolean longMessageMode)
	{
		this(msg, SmsDcs.getGeneralDataCodingDcs(alphabet, messageClass));
		this.setLongMessageMode(longMessageMode);
	}

	/**
	 * Creates an SmsTextMessage with automatic Alphabet and autodetect encoding
	 *
	 * @param msg
	 *            The message
	 */
	public SmsTextMessage(String msg)
	{
		SmsAlphabet smsAlphabet = SmsAlphabet.GSM;
		String encoding = StringUtil.detectEncodingType(msg);
		SmsMsgClass messageClass = SmsMsgClass.CLASS_UNKNOWN;
		System.out.println("Encoding detected: " + encoding);
		if (!encoding.equals("US-ASCII"))
		{
			try
			{
				msg = new String(msg.getBytes(encoding), "UTF-8");
				smsAlphabet = SmsAlphabet.UCS2;
			}
			catch (UnsupportedEncodingException ignore)
			{
			}
		}
		this.setText(msg, SmsDcs.getGeneralDataCodingDcs(smsAlphabet, messageClass));
	}

	/**
	 * Returns the text message.
	 */
	public String getText()
	{
		return text_;
	}

	/**
	 * Sets the text.
	 * 
	 * @param text
	 */
	public void setText(String text)
	{
		if (text == null)
		{
			throw new IllegalArgumentException("Text cannot be null, use an empty string instead.");
		}
		if (dcs_ != null)
		{
			if (dcs_.getAlphabet() == SmsAlphabet.GSM)
			{
				text = Normalizer.normalize(text, Normalizer.Form.NFD);
				text = text.replaceAll("[\\p{InCombiningDiacriticalMarks}]", "");
				text = text.replaceAll("€", "euros");
			}
		}
		text_ = text;
	}

	/**
	 * Sets the text.
	 * 
	 * @param text
	 */
	public void setText(String text, SmsDcs dcs)
	{
		// Check input for null
		if (text == null)
		{
			throw new IllegalArgumentException("text cannot be null, use an empty string instead.");
		}
		if (dcs == null)
		{
			throw new IllegalArgumentException("dcs cannot be null.");
		}
		dcs_ = dcs;
		this.setText(text);
	}

	/**
	 * Returns the dcs.
	 */
	public SmsDcs getDcs()
	{
		return dcs_;
	}

	/**
	 * Returns the user data.
	 * 
	 * @return user data
	 */
	public SmsUserData getUserData()
	{
		SmsUserData ud = null;
		try
		{
			switch (dcs_.getAlphabet())
			{
			case GSM:
				ud = new SmsUserData(SmsPduUtil.getSeptets(text_), text_.length(), dcs_);
				break;
			case LATIN1:
				byte[] bytesUTF8 = text_.getBytes("UTF-8");
				ud = new SmsUserData(bytesUTF8, bytesUTF8.length, dcs_);
				break;
			case UCS2:
				byte[] bytesUTF16 = text_.getBytes("UTF_16BE");
				ud = new SmsUserData(bytesUTF16, bytesUTF16.length, dcs_);
				break;
			default:
				ud = null;
				break;
			}
		}
		catch (Exception e)
		{
		}
		return ud;
	}

	/**
	 * Returns null.
	 */
	public SmsUdhElement[] getUdhElements()
	{
		return null;
	}
}


src/main/java/org/marre/sms/transport/ucp/UcpSeries50.java
/* ***** BEGIN LICENSE BLOCK *****
 * Version: MPL 1.1/GPL 2.0/LGPL 2.1
 *
 * The contents of this file are subject to the Mozilla Public License Version
 * 1.1 (the "License"); you may not use this file except in compliance with
 * the License. You may obtain a copy of the License at
 * http://www.mozilla.org/MPL/
 *
 * Software distributed under the License is distributed on an "AS IS" basis,
 * WITHOUT WARRANTY OF ANY KIND, either express or implied. See the License
 * for the specific language governing rights and limitations under the
 * License.
 *
 * The Original Code is "SMS Library for the Java platform".
 *
 * The Initial Developer of the Original Code is Markus Eriksson.
 * Portions created by the Initial Developer are Copyright (C) 2002
 * the Initial Developer. All Rights Reserved.
 *
 * Contributor(s):
 *
 * Alternatively, the contents of this file may be used under the terms of
 * either the GNU General Public License Version 2 or later (the "GPL"), or
 * the GNU Lesser General Public License Version 2.1 or later (the "LGPL"),
 * in which case the provisions of the GPL or the LGPL are applicable instead
 * of those above. If you wish to allow use of your version of this file only
 * under the terms of either the GPL or the LGPL, and not to allow others to
 * use your version of this file under the terms of the MPL, indicate your
 * decision by deleting the provisions above and replace them with the notice
 * and other provisions required by the GPL or the LGPL. If you do not delete
 * the provisions above, a recipient may use your version of this file under
 * the terms of any one of the MPL, the GPL or the LGPL.
 *
 * ***** END LICENSE BLOCK ***** */
package org.marre.sms.transport.ucp;

import org.marre.util.StringUtil;

/**
 * 
 * @author Markus Eriksson
 * @version $Id$
 */
public class UcpSeries50 extends UcpMsg
{
	public static final byte OP_SUBMIT_SHORT_MESSAGE = 51;
	public static final byte OP_DELIVER_SHORT_MESSAGE = 52;
	public static final byte OP_DELIVER_NOTIFICATION = 53;
	public static final byte OP_MODIFY_MESSAGE = 54;
	public static final byte OP_INQUIRY_MESSAGE = 55;
	public static final byte OP_DELETE_MESSAGE = 56;
	public static final byte OP_RESPONSE_INQUIRY_MESSAGE = 57;
	public static final byte OP_RESPONSE_DELETE_MESSAGE = 58;
	public static final byte XSER_TYPE_UDH = 1;
	public static final byte XSER_TYPE_DCS = 2;
	protected static final int FIELD_ADC = 0;
	protected static final int FIELD_OADC = 1;
	protected static final int FIELD_AC = 2;
	protected static final int FIELD_NRQ = 3;
	protected static final int FIELD_NADC = 4;
	protected static final int FIELD_NT = 5;
	protected static final int FIELD_NPID = 6;
	protected static final int FIELD_LRQ = 7;
	protected static final int FIELD_LRAD = 8;
	protected static final int FIELD_LPID = 9;
	protected static final int FIELD_DD = 10;
	protected static final int FIELD_DDT = 11;
	protected static final int FIELD_VP = 12;
	protected static final int FIELD_RPID = 13;
	protected static final int FIELD_SCTS = 14;
	protected static final int FIELD_DST = 15;
	protected static final int FIELD_RSN = 16;
	protected static final int FIELD_DSCTS = 17;
	protected static final int FIELD_MT = 18;
	protected static final int FIELD_NB = 19;
	protected static final int FIELD_MSG = 20; // NMsg, AMsg or TMsg (MT)
	protected static final int FIELD_MMS = 21;
	protected static final int FIELD_PR = 22;
	protected static final int FIELD_DCS = 23;
	protected static final int FIELD_MCLS = 24;
	protected static final int FIELD_RPI = 25;
	protected static final int FIELD_CPG = 26;
	protected static final int FIELD_RPLY = 27;
	protected static final int FIELD_OTOA = 28;
	protected static final int FIELD_HPLMN = 29;
	protected static final int FIELD_XSER = 30;
	protected static final int FIELD_RES4 = 31;
	protected static final int FIELD_RES5 = 32;

	public UcpSeries50(byte operation)
	{
		super(33);
		setOR('O');
		setOT(operation);
	}

	public void clearXSer()
	{
		ucpFields_[FIELD_XSER] = null;
	}

	public void addXSer(byte type, byte data)
	{
		StringBuffer xSerBuff = new StringBuffer();
		xSerBuff.append(StringUtil.byteToHexString(type));
		xSerBuff.append("01");
		xSerBuff.append(StringUtil.byteToHexString(data));
		ucpFields_[FIELD_XSER] = (ucpFields_[FIELD_XSER] == null) ? (xSerBuff.toString()) : (ucpFields_[FIELD_XSER] + xSerBuff.toString());
	}

	public void addXSer(byte type, byte[] data)
	{
		StringBuffer xSerBuff = new StringBuffer();
		xSerBuff.append(StringUtil.byteToHexString(type));
		xSerBuff.append(StringUtil.byteToHexString((byte) (data.length & 0xff)));
		xSerBuff.append(StringUtil.bytesToHexString(data));
		ucpFields_[FIELD_XSER] = (ucpFields_[FIELD_XSER] == null) ? (xSerBuff.toString()) : (ucpFields_[FIELD_XSER] + xSerBuff.toString());
	}

	public boolean parseResult(String smscresponse)
	{
		boolean result = false;
		String[] resp = smscresponse.split("/");
		if (resp != null && resp.length > 0)
		{
			if (resp[4].equalsIgnoreCase("A"))
			{ // A = ACK, N = NACK
				result = true;
			}
		}
		return result;
	}
}


src/main/java/org/marre/sms/SmsPduUtil.java
/* ***** BEGIN LICENSE BLOCK *****
 * Version: MPL 1.1/GPL 2.0/LGPL 2.1
 *
 * The contents of this file are subject to the Mozilla Public License Version
 * 1.1 (the "License"); you may not use this file except in compliance with
 * the License. You may obtain a copy of the License at
 * http://www.mozilla.org/MPL/
 *
 * Software distributed under the License is distributed on an "AS IS" basis,
 * WITHOUT WARRANTY OF ANY KIND, either express or implied. See the License
 * for the specific language governing rights and limitations under the
 * License.
 *
 * The Original Code is "SMS Library for the Java platform".
 *
 * The Initial Developer of the Original Code is Markus Eriksson.
 * Portions created by the Initial Developer are Copyright (C) 2002
 * the Initial Developer. All Rights Reserved.
 *
 * Contributor(s):
 *   Boris von Loesch
 *
 * Alternatively, the contents of this file may be used under the terms of
 * either the GNU General Public License Version 2 or later (the "GPL"), or
 * the GNU Lesser General Public License Version 2.1 or later (the "LGPL"),
 * in which case the provisions of the GPL or the LGPL are applicable instead
 * of those above. If you wish to allow use of your version of this file only
 * under the terms of either the GPL or the LGPL, and not to allow others to
 * use your version of this file under the terms of the MPL, indicate your
 * decision by deleting the provisions above and replace them with the notice
 * and other provisions required by the GPL or the LGPL. If you do not delete
 * the provisions above, a recipient may use your version of this file under
 * the terms of any one of the MPL, the GPL or the LGPL.
 *
 * ***** END LICENSE BLOCK ***** */
package org.marre.sms;

import java.io.ByteArrayInputStream;
import java.io.ByteArrayOutputStream;
import java.io.IOException;
import java.io.InputStream;
import java.io.OutputStream;

/**
 * Various functions to encode and decode strings
 * 
 * @author Markus Eriksson
 */
public final class SmsPduUtil
{
	public static final char EXT_TABLE_PREFIX = 0x1B;
	/**
	 * Default alphabet table according to GSM 03.38.
	 * 
	 * See http://www.unicode.org/Public/MAPPINGS/ETSI/GSM0338.TXT
	 */
	public static final char[] GSM_DEFAULT_ALPHABET_TABLE = {
			// 0 '@', '?', '$', '?', '?', '?', '?', '?',
			'@', 163, '$', 165, 232, 233, 249, 236,
			// 8 '?', '?', LF, '?', '?', CR, '?', '?',
			242, 199, 10, 216, 248, 13, 197, 229,
			// 16 'delta', '_', 'phi', 'gamma', 'lambda', 'omega', 'pi', 'psi',
			0x394, '_', 0x3a6, 0x393, 0x39b, 0x3a9, 0x3a0, 0x3a8,
			// 24 'sigma', 'theta', 'xi', 'EXT', '?', '?', '?', '?',
			0x3a3, 0x398, 0x39e, 0xa0, 198, 230, 223, 201,
			// 32 ' ', '!', '"', '#', '?', '%', '&', ''',
			' ', '!', '"', '#', 164, '%', '&', '\'',
			// 40 '(', ')', '*', '+', ',', '-', '.', '/',
			'(', ')', '*', '+', ',', '-', '.', '/',
			// 48 '0', '1', '2', '3', '4', '5', '6', '7',
			'0', '1', '2', '3', '4', '5', '6', '7',
			// 56 '8', '9', ':', ';', '<', '=', '>', '?',
			'8', '9', ':', ';', '<', '=', '>', '?',
			// 64 '?', 'A', 'B', 'C', 'D', 'E', 'F', 'G',
			161, 'A', 'B', 'C', 'D', 'E', 'F', 'G',
			// 72 'H', 'I', 'J', 'K', 'L', 'M', 'N', 'O',
			'H', 'I', 'J', 'K', 'L', 'M', 'N', 'O',
			// 80 'P', 'Q', 'R', 'S', 'T', 'U', 'V', 'W',
			'P', 'Q', 'R', 'S', 'T', 'U', 'V', 'W',
			// 88 'X', 'Y', 'Z', '?', '?', '?', '?', '?',
			'X', 'Y', 'Z', 196, 214, 209, 220, 167,
			// 96 '?', 'a', 'b', 'c', 'd', 'e', 'f', 'g',
			191, 'a', 'b', 'c', 'd', 'e', 'f', 'g',
			// 104 'h', 'i', 'j', 'k', 'l', 'm', 'n', 'o',
			'h', 'i', 'j', 'k', 'l', 'm', 'n', 'o',
			// 112 'p', 'q', 'r', 's', 't', 'u', 'v', 'w',
			'p', 'q', 'r', 's', 't', 'u', 'v', 'w',
			// 120 'x', 'y', 'z', '?', '?', '?', '?', '?',
			'x', 'y', 'z', 228, 246, 241, 252, 224 };
	/**
	 * Some alternative character encodings.
	 * 
	 * The table is encoded as pairs with unicode value and gsm charset value. <br>
	 * Ex:
	 * 
	 * <pre>
	 * char unicode = GSM_DEFAULT_ALPHABET_ALTERNATIVES[i * 2];
	 * char gsm = GSM_DEFAULT_ALPHABET_ALTERNATIVES[i * 2 + 1];
	 * </pre>
	 * 
	 * See http://www.unicode.org/Public/MAPPINGS/ETSI/GSM0338.TXT
	 */
	public static final char[] GSM_DEFAULT_ALPHABET_ALTERNATIVES = {
			// LATIN CAPITAL LETTER C WITH CEDILLA (see note above)
			0x00c7, 0x09,
			// GREEK CAPITAL LETTER ALPHA
			0x0391, 0x41,
			// GREEK CAPITAL LETTER BETA
			0x0392, 0x42,
			// GREEK CAPITAL LETTER ETA
			0x0397, 0x48,
			// GREEK CAPITAL LETTER IOTA
			0x0399, 0x49,
			// GREEK CAPITAL LETTER KAPPA
			0x039a, 0x4b,
			// GREEK CAPITAL LETTER MU
			0x039c, 0x4d,
			// GREEK CAPITAL LETTER NU
			0x039d, 0x4e,
			// GREEK CAPITAL LETTER OMICRON
			0x039f, 0x4f,
			// GREEK CAPITAL LETTER RHO
			0x03a1, 0x50,
			// GREEK CAPITAL LETTER TAU
			0x03a4, 0x54,
			// GREEK CAPITAL LETTER UPSILON
			0x03a5, 0x55,
			// GREEK CAPITAL LETTER CHI
			0x03a7, 0x58,
			// GREEK CAPITAL LETTER ZETA
			0x0396, 0x5a };

	/**
	 * This class isn't intended to be instantiated
	 */
	private SmsPduUtil()
	{
	}

	/**
	 * Pack the given string into septets
	 * 
	 */
	public static byte[] getSeptets(String msg)
	{
		ByteArrayOutputStream baos = new ByteArrayOutputStream(140);
		try
		{
			writeSeptets(baos, msg);
			baos.close();
		}
		catch (IOException ex)
		{
			// Should not happen...
			throw new RuntimeException(ex);
		}
		return baos.toByteArray();
	}

	/**
	 * Pack the given string into septets.
	 * 
	 * @param os
	 *            Write the septets into this stream
	 * @param msg
	 *            The message to encode
	 * @throws IOException
	 *             Thrown when failing to write to os
	 */
	public static void writeSeptets(OutputStream os, String msg) throws IOException
	{
		int data = 0;
		int nBits = 0;
		for (int i = 0; i < msg.length(); i++)
		{
			byte gsmChar = toGsmCharset(msg.charAt(i));
			data |= (gsmChar << nBits);
			nBits += 7;
			while (nBits >= 8)
			{
				os.write((char) (data & 0xff));
				data >>>= 8;
				nBits -= 8;
			} // while
		} // for
			// Write remaining byte
		if (nBits > 0)
		{
			os.write(data);
		}
	}

	/**
	 * Decodes a 7-bit encoded string from the given byte array
	 * 
	 * @param data
	 *            The byte array to read from
	 * @param length
	 *            Number of decoded chars to read from the stream
	 * @return The decoded string
	 */
	public static String readSeptets(byte[] data, int length)
	{
		if (data == null)
		{
			return null;
		}
		try
		{
			return readSeptets(new ByteArrayInputStream(data), length);
		}
		catch (IOException ex)
		{
			// Shouldn't happen since we are reading from a bytearray...
			return null;
		}
	}

	/**
	 * Decodes a 7-bit encoded string from the stream
	 * 
	 * @param is
	 *            The stream to read from
	 * @param length
	 *            Number of decoded chars to read from the stream
	 * @return The decoded string
	 * @throws IOException
	 *             when failing to read from is
	 */
	public static String readSeptets(InputStream is, int length) throws IOException
	{
		StringBuilder msg = new StringBuilder(160);
		int rest = 0;
		int restBits = 0;
		while (msg.length() < length)
		{
			int data = is.read();
			if (data == -1)
			{
				throw new IOException("Unexpected end of stream");
			}
			rest |= (data << restBits);
			restBits += 8;
			while ((msg.length() < length) && (restBits >= 7))
			{
				msg.append(fromGsmCharset((byte) (rest & 0x7f)));
				rest >>>= 7;
				restBits -= 7;
			}
		} // for
		return msg.toString();
	}

	/**
	 * Writes the given phonenumber to the stream (BCD coded)
	 * 
	 * @param os
	 *            Stream to write to
	 * @param number
	 *            Number to convert
	 * @throws IOException
	 *             when failing to write to os
	 */
	public static void writeBcdNumber(OutputStream os, String number) throws IOException
	{
		int bcd = 0x00;
		int n = 0;
		// First convert to a "half octet" value
		for (int i = 0; i < number.length(); i++)
		{
			switch (number.charAt(i))
			{
			case '0':
				bcd |= 0x00;
				break;
			case '1':
				bcd |= 0x10;
				break;
			case '2':
				bcd |= 0x20;
				break;
			case '3':
				bcd |= 0x30;
				break;
			case '4':
				bcd |= 0x40;
				break;
			case '5':
				bcd |= 0x50;
				break;
			case '6':
				bcd |= 0x60;
				break;
			case '7':
				bcd |= 0x70;
				break;
			case '8':
				bcd |= 0x80;
				break;
			case '9':
				bcd |= 0x90;
				break;
			case '*':
				bcd |= 0xA0;
				break;
			case '#':
				bcd |= 0xB0;
				break;
			case 'a':
				bcd |= 0xC0;
				break;
			case 'b':
				bcd |= 0xE0;
				break;
			}
			n++;
			if (n == 2)
			{
				os.write(bcd);
				n = 0;
				bcd = 0x00;
			}
			else
			{
				bcd >>= 4;
			}
		}
		if (n == 1)
		{
			bcd |= 0xF0;
			os.write(bcd);
		}
	}

	/**
	 * Converts bytes to BCD format
	 * 
	 * @param is
	 *            The byte InputStream
	 * @param length
	 *            how many
	 * @return Decoded number
	 */
	public static String readBcdNumber(InputStream is, int length) throws IOException
	{
		byte[] arr = new byte[length];
		is.read(arr, 0, length);
		return readBcdNumber(arr, 0, length);
	}

	/**
	 * Converts bytes to BCD format
	 * 
	 * @param data
	 *            bytearray
	 * @param length
	 *            how many
	 * @param offset
	 * @return Decoded number
	 */
	public static String readBcdNumber(byte[] data, int offset, int length)
	{
		StringBuilder out = new StringBuilder();
		for (int i = offset; i < offset + length; i++)
		{
			int arrb = data[i];
			if ((data[i] & 15) <= 9)
			{
				out.append("").append(data[i] & 15);
			}
			if ((data[i] & 15) == 0xA)
			{
				out.append("*");
			}
			if ((data[i] & 15) == 0xB)
			{
				out.append("#");
			}
			arrb = (arrb >>> 4);
			if ((arrb & 15) <= 9)
			{
				out.append("").append(arrb & 15);
			}
			if ((arrb & 15) == 0xA)
			{
				out.append("*");
			}
			if ((arrb & 15) == 0xB)
			{
				out.append("#");
			}
		}
		return out.toString();
	}

	/**
	 * Convert from the GSM charset to a unicode char
	 * 
	 * @param gsmChar
	 *            The gsm char to convert
	 * @return Unicode representation of the given gsm char
	 */
	public static char fromGsmCharset(byte gsmChar)
	{
		return GSM_DEFAULT_ALPHABET_TABLE[gsmChar];
	}

	/**
	 * Converts a unicode string to GSM charset
	 * 
	 * @param str
	 *            String to convert
	 * @return The string GSM encoded
	 */
	public static byte[] toGsmCharset(String str)
	{
		byte[] gsmBytes = new byte[str.length()];
		for (int i = 0; i < gsmBytes.length; i++)
		{
			gsmBytes[i] = toGsmCharset(str.charAt(i));
		}
		return gsmBytes;
	}

	/**
	 * Convert a unicode char to a GSM char
	 * 
	 * @param ch
	 *            The unicode char to convert
	 * @return GSM representation of the given unicode char
	 */
	public static byte toGsmCharset(char ch)
	{
		// First check through the GSM charset table
		for (int i = 0; i < GSM_DEFAULT_ALPHABET_TABLE.length; i++)
		{
			if (GSM_DEFAULT_ALPHABET_TABLE[i] == ch)
			{
				// Found the correct char
				return (byte) i;
			}
		}
		// Alternative chars.
		for (int i = 0; i < GSM_DEFAULT_ALPHABET_ALTERNATIVES.length / 2; i += 2)
		{
			if (GSM_DEFAULT_ALPHABET_ALTERNATIVES[i * 2] == ch)
			{
				return (byte) (GSM_DEFAULT_ALPHABET_ALTERNATIVES[i * 2 + 1] & 0x7f);
			}
		}
		// Couldn't find a valid char
		return '?';
	}

	/**
	 * 
	 * @param src
	 * @param srcStart
	 * @param dest
	 * @param destStart
	 * @param destBitOffset
	 * @param lengthInBits
	 *            In bits
	 */
	public static void arrayCopy(byte[] src, int srcStart, byte[] dest, int destStart, int destBitOffset, int lengthInBits)
	{
		int c = 0;
		int nBytes = lengthInBits / 8;
		int nRestBits = lengthInBits % 8;
		for (int i = 0; i < nBytes; i++)
		{
			c |= ((src[srcStart + i] & 0xff) << destBitOffset);
			dest[destStart + i] |= (byte) (c & 0xff);
			c >>>= 8;
		}
		if (nRestBits > 0)
		{
			c |= ((src[srcStart + nBytes] & (0xff >> (8 - nRestBits))) << destBitOffset);
		}
		if ((nRestBits + destBitOffset) > 0)
		{
			dest[destStart + nBytes] |= c & 0xff;
		}
	}
}
